pkg/agent/clean/adunmigration/migrate.go

```diff
@@ -0,0 +1,499 @@
+			logrus.Infof("[%v] ad-guid migration has already been completed. To clean-up missing users, you can run the utility manually", migrateAdUserOperation)
+			return nil
+		case activedirectory.StatusMigrationFinishedWithSkipped:
+			logrus.Infof("[%v] ad-guid migration has already been completed. To try and resolve skipped users, you can run the utility manually", migrateAdUserOperation)
+			return nil
+		}
+
+	}
+	return UnmigrateAdGUIDUsers(&sc.RESTConfig, false, false)
+}
+
+// UnmigrateAdGUIDUsers will cycle through all users, ctrb, ptrb, tokens and migrate them to an
+// appropriate DN-based PrincipalID.
+func UnmigrateAdGUIDUsers(clientConfig *restclient.Config, dryRun bool, deleteMissingUsers bool) error {
+	if dryRun {
+		logrus.Infof("[%v] dryRun is true, no objects will be deleted/modified", migrateAdUserOperation)
+		deleteMissingUsers = false
+	} else if deleteMissingUsers {
+		logrus.Infof("[%v] deleteMissingUsers is true, GUID-based users not present in Active Directory will be deleted", migrateAdUserOperation)
+	}
+
+	sc, adConfig, err := prepareClientContexts(clientConfig)
+	if err != nil {
+		return err
+	}
+
+	migrationConfigMap, err := sc.Core.ConfigMaps(activedirectory.StatusConfigMapNamespace).GetNamespaced(activedirectory.StatusConfigMapNamespace, activedirectory.StatusConfigMapName, metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		logrus.Errorf("[%v] unable to check unmigration configmap: %v", migrateAdUserOperation, err)
+		logrus.Errorf("[%v] cannot determine if it is safe to proceed. refusing to run", migrateAdUserOperation)
+		return nil
+	}
+	if migrationConfigMap != nil {
+		migrationStatus := migrationConfigMap.Data[activedirectory.StatusMigrationField]
+		switch migrationStatus {
+		case activedirectory.StatusMigrationRunning:
+			logrus.Infof("[%v] ad-guid migration is currently running, refusing to run again concurrently", migrateAdUserOperation)
+			return nil
+		}
+	}
+
+	finalStatus := activedirectory.StatusMigrationFinished
+
+	// set the status to running and reset the unmigrated fields
+	if !dryRun {
+		err = updateMigrationStatus(sc, activedirectory.StatusMigrationField, activedirectory.StatusMigrationRunning)
+		if err != nil {
+			return fmt.Errorf("unable to update migration status configmap: %v", err)
+		}
+		updateUnmigratedUsers("", migrateStatusSkipped, true, sc)
+		updateUnmigratedUsers("", migrateStatusMissing, true, sc)
+		// If we return past this point, no matter how we got there, make sure we update the configmap to clear the
+		// status away from "running." If we fail to do this, we block AD-based logins indefinitely.
+		defer func(sc *config.ScaledContext, status string, value string) {
+			err := updateMigrationStatus(sc, status, value)
+			if err != nil {
+				logrus.Errorf("[%v] unable to update migration status configmap: %v", migrateAdUserOperation, err)
+			}
+		}(sc, activedirectory.StatusMigrationField, finalStatus)
```

Comment:
This `defer` is going to set the `value`/`finalStatus` to the value at the time the defer is created rather than at the time the function is run.

Compare [this defer example](https://go.dev/play/p/Pajyo6PgVZ-) with this [other defer example](https://go.dev/play/p/TFldgqAlO6F).

The `finalStatus` changes that show up on lines like 181, 186, etc are not taken into account when the defer happens.

Shouldn't this be changed?